### PR TITLE
Improve EbbAndFlow scenario

### PR DESCRIPTION
**Pull request by stephanos**
## What was changed

- added resume capability
- added visibility check for failed workflows
- added integration test

## Why?

Make it more robust for regular CICD runs.

## Checklist

1. Closes 

2. How was this tested:

3. Any docs updates needed?

### CODE REVIEW

File: cmd/cmdoptions/logging.go [hunk 1]
```diff
@@ -16,13 +16,20 @@ type LoggingOptions struct {
 	LogLevel string
 	// Log encoding (console json)
 	LogEncoding string
+	// Prepared logger to use instead of creating a new one from options. Progamatic use only.
+	PreparedLogger *zap.SugaredLogger
```
Comment (stephanos) [inline cmd/cmdoptions/logging.go:20]: Used to pass in custom logger for testing.

File: loadgen/ebbandflow/ebb_and_flow.go [hunk 2]
```diff
@@ -15,7 +15,7 @@ type WorkflowOutput struct {
 }
 
 type ActivityTiming struct {
-	FairnessKey       string        `json:"fairnessKey"`
-	FairnessWeight    float32       `json:"fairnessWeight"`
-	ScheduleToStartMS time.Duration `json:"scheduleToStartMS"`
+	FairnessKey     string        `json:"k"`
+	FairnessWeight  float32       `json:"w"`
+	ScheduleToStart time.Duration `json:"d"`
```
Comment (stephanos) [inline loadgen/ebbandflow/ebb_and_flow.go:20]: ^ shortened JSON keys to reduce payload size.

File: loadgen/helpers.go [hunk 3]
```diff
@@ -101,3 +102,33 @@ func MinVisibilityCountEventually(
 
 	return nil
 }
+
+// VerifyNoFailedWorkflows verifies that there are no failed or terminated workflows for the given search attribute.
+func VerifyNoFailedWorkflows(ctx context.Context, info ScenarioInfo, searchAttribute, runID string) error {
```
Comment (stephanos) [inline loadgen/helpers.go:107]: New re-usable helper for scenarios.

File: scenarios/ebb_and_flow.go [hunk 4]
```diff
@@ -242,14 +242,9 @@ func (e *ebbAndFlow) run(ctx context.Context) error {
...
 
-	// Post-scenario: verify that at least one iteration was completed.
-	completedWorkflows := int(e.processedCount.Load())
-	if completedWorkflows == 0 {
-		return fmt.Errorf("no workflows completed")
-	}
-
 	// Post-scenario: verify reported workflow completion count from Visibility.
-	return loadgen.MinVisibilityCountEventually(
+	completedWorkflows := max(int(e.processedCount.Load()), 1)
```
Comment (stephanos) [inline scenarios/ebb_and_flow.go:246]: Assume we have at least one in case we resume.

File: scenarios/ebb_and_flow.go [hunk 5]
```diff
@@ -259,7 +254,12 @@ func (e *ebbAndFlow) run(ctx context.Context) error {
 		},
 		completedWorkflows,
 		config.VisibilityVerificationTimeout,
-	)
+	); err != nil {
+		return err
+	}
+
+	// Post-scenario: ensure there are no failed or terminated workflows for this run.
+	return loadgen.VerifyNoFailedWorkflows(ctx, e.ScenarioInfo, EbbAndFlowScenarioIdSearchAttribute, e.ScenarioInfo.RunID)
```
Comment (stephanos) [inline scenarios/ebb_and_flow.go:262]: New addition.

File: scenarios/throughput_stress_test.go
Comment (stephanos) [inline scenarios/throughput_stress_test.go:1]: Just some cosmetic changes ...

File: scenarios/throughput_stress_test.go [hunk 6]
```diff
@@ -14,7 +14,6 @@ import (
 func TestThroughputStress(t *testing.T) {
```
Comment (stephanos) [inline scenarios/throughput_stress_test.go:14]: Made some cosmetic changes ...

File: workers/test_env.go [hunk 7]
```diff
@@ -81,16 +86,15 @@ func (env *TestEnvironment) NexusEndpointName() string {
 }
 
 func SetupTestEnvironment(t *testing.T, opts ...TestEnvOption) *TestEnvironment {
-	logger := zaptest.NewLogger(t).Sugar()
```
Comment (stephanos) [inline workers/test_env.go:84]: Moved the logger into `RunExecutorTest` so it's per test execution instead of per test setup which makes sure the logs are associated with the right test now.

File: scenarios/ebb_and_flow.go [hunk 8]
```diff
@@ -53,11 +53,28 @@ type ebbAndFlowConfig struct {
...
+	config             *ebbAndFlowConfig
+	rng                *rand.Rand
+	id                 string
+	isResuming         bool
+	startTime          time.Time
+	startedWorkflows   atomic.Int64
+	completedWorkflows atomic.Int64
+	fairnessTracker    *ebbandflow.FairnessTracker
+	stateLock          sync.Mutex
+	state              *ebbAndFlowState
```
Comment (stephanos) [inline scenarios/ebb_and_flow.go:74]: Merged `ebbAndFlow` into here; simpler.

File: scenarios/ebb_and_flow.go [hunk 9]
```diff
@@ -53,11 +53,28 @@ type ebbAndFlowConfig struct {
 	SleepActivityConfig           *loadgen.SleepActivityConfig
 }
 
+type ebbAndFlowState struct {
```
Comment (stephanos) [inline scenarios/ebb_and_flow.go:57]: New state to be picked up on resume again.

File: scenarios/ebb_and_flow_test.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
+package scenarios
```
Comment (stephanos) [inline scenarios/ebb_and_flow_test.go:1]: New new test verifying some resume edge cases.

File: scenarios/ebb_and_flow.go [hunk 11]
```diff
@@ -233,36 +228,80 @@ func (e *ebbAndFlow) run(ctx context.Context) error {
...
-	// Post-scenario: verify that at least one iteration was completed.
-	completedWorkflows := int(e.processedCount.Load())
-	if completedWorkflows == 0 {
-		return fmt.Errorf("no workflows completed")
+	e.stateLock.Lock()
+	totalCompletedWorkflows := int(e.state.TotalCompletedWorkflows)
+	e.stateLock.Unlock()
+
+	// Post-scenario: verify that at least one workflow was completed.
+	if totalCompletedWorkflows == 0 {
```
Comment (stephanos) [inline scenarios/ebb_and_flow.go:256]: This is now using the resume-aware counter; so it's not going to be zero if we crash and resume after the iterations were all done.

File: scenarios/ebb_and_flow.go [hunk 12]
```diff
@@ -233,36 +228,80 @@ func (e *ebbAndFlow) run(ctx context.Context) error {
...
-		completedWorkflows,
+		totalCompletedWorkflows,
 		config.VisibilityVerificationTimeout,
-	)
+	); err != nil {
+		return err
+	}
+
+	// Post-scenario: ensure there are no failed or terminated workflows for this run.
+	return loadgen.VerifyNoFailedWorkflows(ctx, e.ScenarioInfo, EbbAndFlowScenarioIdSearchAttribute, e.ScenarioInfo.RunID)
```
Comment (stephanos) [inline scenarios/ebb_and_flow.go:276]: Net new assertion.

File: scenarios/ebb_and_flow.go [hunk 13]
```diff
@@ -233,36 +228,80 @@ func (e *ebbAndFlow) run(ctx context.Context) error {
 				iter++
 			}
 		case <-backlogTicker.C:
-			e.Logger.Infof("Backlog: %d, target: %d, rate: %d/s, gen: %d, proc: %d",
-				backlog, target, rate, generated, processed)
+			e.Logger.Debugf("Backlog: %d, target: %d, rate: %d/s, started: %d, completed: %d",
```
Comment (stephanos) [inline scenarios/ebb_and_flow.go:242]: Less noisy.

File: workers/test_env.go [hunk 14]
```diff
@@ -27,14 +30,6 @@ const (
 	workerShutdownTimeout = 5 * time.Second
 )
 
-var (
-	testStart          = time.Now().Unix()
-	workerMutex        sync.RWMutex
-	workerBuildOnce    = map[cmdoptions.Language]*sync.Once{}
-	workerBuildErrs    = map[cmdoptions.Language]error{}
-	workerCleanupFuncs []func()
```
Comment (stephanos) [inline workers/test_env.go:35]: scoped all these per test env to prevent races between test envs.

File: workers/go/ebbandflow/activities.go [hunk 15]
```diff
@@ -67,7 +67,7 @@ func (a Activities) ProcessFairnessReport(ctx context.Context, report ebbandflow
 				offender.ViolationSeverity)
 			errorFields = append(errorFields, fmt.Sprintf("topViolator%d", i+1), violatorSummary)
 		}
-		logger.Error("Fairness violation", errorFields...)
+		logger.Error("Fairness status: violated", errorFields...)
```
Comment (Sushisource) [inline workers/go/ebbandflow/activities.go:70]: 😆 I love this

File: workers/test_env.go [hunk 16]
```diff
@@ -57,15 +52,23 @@ func WithNexusEndpoint(taskQueue string) TestEnvOption {
...
+	devServer          *testsuite.DevServer
+	temporalClient     client.Client
+	baseDir            string
+	buildDir           string
+	repoDir            string
+	nexusEndpointName  string
+	workerMutex        sync.RWMutex
+	workerBuildOnce    map[cmdoptions.Language]*sync.Once
+	workerBuildErrs    map[cmdoptions.Language]error
+	workerCleanupFuncs []func()
```
Comment (Sushisource) [inline workers/test_env.go]: nit: Could be its own struct so the locking is more encapsulated

**Review by Sushisource — APPROVED** [review]
Looking good to me. Only one nonblocking suggestion.

File: workers/test_env.go (hunk 16 above)
Comment (stephanos) [reply workers/test_env.go]: Good call; it's actually much better now.
